### PR TITLE
Implementing all 4 typical "messages" as in proftpd and other commercial FTP servers

### PR DESCRIPTION
**Pull request by syncplify**
The FTP protocol allows for (and many open source as well as commercial FTP servers support) custom messages to be sent to the client in 4 specific moments of the session:

1. upon client connection
2. upon successful user auth
3. upon failed user auth
4. upon receipt of the QUIT command (just prior to closing the connection)

We noticed that ftpserverlib only supports point number 1, therefore we would like to contribute some code to implement the other 3.

Our notes:

- Implementing point number 4 was quite straightforward by means of an "extended" interface (like you did with several other things)
- You may not like our implementation of point number 2 and 3 here above, because it breaks backwards compatibility by changing the AuthUser method prototype; this was the easiest way to implement it, but if you want to do it in a more backward-compatible way, by all means, feel free to totally reject our implementation

Regardless of the implementation details, we see there's value to this functionality, so we hope you'll consider including it into your excellent library.

Thank you,
kind regards.

### CODE REVIEW

**Comment by drakkan** [comment]
Hello,

thank you for this contribution. 
I took a quick look at your patch. Do you need to set different messages based on the user?
If the messages are global to the server maybe you can just add some new settings and thus keep backward compatibility. Not sure if I missing something

**Comment by syncplify** [comment]
Having those messages returned based on the user can be very useful indeed. Some enterprise customers may use variables or even need to return a totally different message based upon who's logging in, hence the need to build those messages in real-time, in code, based upon the user.
Of course I do understand the reluctance for breaking backward compatibility, I promise I don't take this lightly. If there is a way to do it without breaking backwards compatibility I am all for it.
I am simply reinforcing that the functionality - in and by itself - would be extremely useful to have in a user-dependent fashion.

File: driver.go [hunk 1]
```diff
@@ -23,7 +23,7 @@ type MainDriver interface {
 	ClientDisconnected(cc ClientContext)
 
 	// AuthUser authenticates the user and selects an handling driver
-	AuthUser(cc ClientContext, user, pass string) (ClientDriver, error)
+	AuthUser(cc ClientContext, user, pass string) (ClientDriver, string, error)
```
Comment (drakkan) [inline driver.go]: I don't have time to test it right now. Is it possible to do this by extending the ClientDriver interface?

File: handle_auth.go [hunk 2]
```diff
@@ -57,16 +57,23 @@ func (c *clientHandler) handleUSER(param string) error {
...
 	var err error
-	c.driver, err = c.server.driver.AuthUser(c, c.user, param)
+	var msg string
+	c.driver, msg, err = c.server.driver.AuthUser(c, c.user, param)
 
 	switch {
 	case err == nil:
-		c.writeMessage(StatusUserLoggedIn, "Password ok, continue")
+		if msg == "" {
+			msg = "Ok"
```
Comment (drakkan) [inline handle_auth.go]: please keep the same messages as before by default

File: handle_misc.go [hunk 3]
```diff
@@ -274,7 +274,14 @@ func (c *clientHandler) handleTYPE(param string) error {
 
 func (c *clientHandler) handleQUIT(param string) error {
 	c.transferWg.Wait()
-	c.writeMessage(StatusClosingControlConn, "Goodbye")
+	msg := "Goodbye"
+	if quitter, ok := c.server.driver.(MainDriverExtensionQuitter); ok {
+		msg = quitter.GetQuitMessage()
+		if msg == "" {
```
Comment (drakkan) [inline handle_misc.go]: I'm not sure if this check is really necessary, if the extension is implemented it should return a non-empty message

File: driver.go (hunk 1 above)
Comment (syncplify) [reply driver.go]: I am back from an extended work trip, I will look into this in the coming days, thank you

File: handle_auth.go (hunk 2 above)
Comment (syncplify) [reply handle_auth.go]: Not a problem, I will change this as per your instructions

File: driver.go (hunk 1 above)
Comment (syncplify) [reply driver.go]: In my last commit (a few minutes ago) I refactored this and implemented this functionality by means of the MainDriverExtensionPostAuthMessage interface; this restores AuthUser to its previous declaration thus preserving backward compatibility

**Comment by syncplify** [comment]
Guys, I've seen the latest codecov reports, but I am not sure they are correct (unless I am missing something obvious, which is entirely possible). In fact, I have added both of my changes to the library's test suite, so I am not sure why codecov now says that those lines are "not covered by tests" (when indeed they are). Would it be possible for you to take a quick look and let me know what I'm missing (if anything)?
Thanks!

File: driver.go [hunk 4]
```diff
@@ -58,6 +58,19 @@ type MainDriverExtensionUserVerifier interface {
 	PreAuthUser(cc ClientContext, user string) error
 }
 
+// MainDriverExtensionPostAuthMessage is an extension that allows to send a message
+// after the authentication
+type MainDriverExtensionPostAuthMessage interface {
+	// PostAuthMessage is called after the authentication
+	GetPostAuthMessage(cc ClientContext, user string, authErr error) string
```
Comment (drakkan) [inline driver.go]: I would have called the new methods `PostAuthMessage`, `QuitMessage` without the `Get` prefix, but this is just a matter of personal preference, feel free to leave the prefix if you prefer

File: handle_auth.go [hunk 5]
```diff
@@ -57,16 +57,30 @@ func (c *clientHandler) handleUSER(param string) error {
...
+	dpa, ok := c.server.driver.(MainDriverExtensionPostAuthMessage)
+	if ok {
+		msg = dpa.GetPostAuthMessage(c, c.user, err)
+	}
+
 	switch {
 	case err == nil:
-		c.writeMessage(StatusUserLoggedIn, "Password ok, continue")
+		if msg == "" {
+			msg = "Password ok, continue"
```
Comment (drakkan) [inline handle_auth.go:71]: You have added `GetPostAuthMessage` to `TestServerDriver` so this line of code is not reached in tests, `msg` is always != ""

File: handle_misc.go [hunk 6]
```diff
@@ -274,7 +274,14 @@ func (c *clientHandler) handleTYPE(param string) error {
 
 func (c *clientHandler) handleQUIT(param string) error {
 	c.transferWg.Wait()
-	c.writeMessage(StatusClosingControlConn, "Goodbye")
+
+	msg := "Goodbye"
+
+	if quitter, ok := c.server.driver.(MainDriverExtensionQuitter); ok {
+		msg = quitter.GetQuitMessage()
```
Comment (drakkan) [inline handle_misc.go]: `TestServerDriver` does not implement the `MainDriverExtensionQuitter` extension so this line of code is not reached in test cases

**Review by drakkan — COMMENTED** [review]
Thank you for working on these additions and sorry for the delay. 

The changes looks good! Please see my minor comments in the review. 
I would like to see a test case checking that the received messages match the expected ones. 

If you want to cover both the old and new code in test case you may evaluate to extend `TestServerDriver` and add the new methods there, so you can use the old driver to cover the old code paths and the new driver to test the new methods. 
I think this requires a fair amount of effort, maybe you can wait for the project owner to tell you if it's needed for merging these changes

File: driver.go (hunk 4 above)
Comment (syncplify) [reply driver.go]: Agreed we will happily change this.

File: handle_misc.go (hunk 6 above)
Comment (syncplify) [reply handle_misc.go]: True, I missed this one. I'll fix it.

File: handle_auth.go (hunk 5 above)
Comment (syncplify) [reply handle_auth.go:71]: Ok, this one will require a little bit of trickery to test both cases (with and without)... but I can work on it.

Comment (syncplify) [reply handle_auth.go:71]: As suggested I will wait for the project owner's feedback on this one... thanks!

**Comment by syncplify** [comment]
I have also fixed the generic HASH function... it was using **strings.SplitN** which fails if the path contains spaces, and substituted it with a custom **advSplitN** func that correctly splits strings in case there are quoted substrings that contain spaces (as the FTP protocol RFC requires).

**Comment by drakkan** [comment]
> I have also fixed the generic HASH function... it was using **strings.SplitN** which fails if the path contains spaces, and substituted it with a custom **advSplitN** func that correctly splits strings in case there are quoted substrings that contain spaces (as the FTP protocol RFC requires).

Thanks for this fix. I think it should be a separate PR and writing a test cases should be quite simple.

As for the original issue, I have added to my TODO to write the required test cases, but I'm very busy at the moment, it might take some time

File: handle_files.go [hunk 7]
```diff
@@ -593,7 +593,12 @@ func (c *clientHandler) handleGenericHash(param string, algo HASHAlgo, isCustomM
 		return nil
 	}
 
-	args := strings.SplitN(param, " ", 3)
+	args, err := advSplitN(param, ' ', 3)
+
+	if err != nil || len(args) < 1 {
+		c.writeMessage(StatusSyntaxErrorParameters, fmt.Sprintf("invalid HASH parameters: %v", param))
```
Comment (drakkan) [inline handle_files.go:599]: missing return

**Comment by fclairamb** [comment]
This was done in #435.

**Pull request closed without merging**